### PR TITLE
fix #484 -- broaden pooling method signatures

### LINKED ISSUE #484: ReverseDiff cannot differentiate `maxpool`
**Edit**
I originally thought this to be an issue with Lux. That's not the case.
Please see [this message](https://github.com/FluxML/NNlib.jl/issues/484#issuecomment-1490418215) for a brief example where ReverseDiff fails to take the gradient of a pooling operation.

---
**Original Description**

ReverseDiff cannot take derivatives of simple models involving Pooling Layers with respect to model parameters.

In the following example, two problems occur:
* if the model is a Pooling Layer only, then no gradient can be computed, but this is kind of understandable, since there are no parameters.
* for a model with more parameters, we get a MethodError for `maxpool!` (see stacktrace below).
I believe it is due to the strict type restrictions in the [method signatures](https://github.com/FluxML/NNlib.jl/blob/5a1c42c8f23b0e45cea77714b84fd231e25897aa/src/pooling.jl#L71) and the way ReverseDiff handles [`similar`](https://github.com/JuliaDiff/ReverseDiff.jl/blob/f06b776333dfd3b5259a27bdfb789d8f017c1ee1/src/tracked.jl#L398)

Example for Julia 1.8.5, 
```
  [d360d2e6] ChainRulesCore v1.15.7
  [b0b7db55] ComponentArrays v0.13.8
  [b2108857] Lux v0.4.48
  [872c559c] NNlib v0.8.19
  [37e2e3b7] ReverseDiff v1.14.4
  [e88e6eb3] Zygote v0.6.59
  [9a3f8284] Random
  [8dfed614] Test
```
Heres the script:
```julia
import Lux
import NNlib
import Zygote
import ReverseDiff
import Random
import ComponentArrays
import Test: @test, @test_throws

rng = Random.default_rng()

x = rand(2,2,1,1)

m = Lux.MaxPool((2,2))
ps, st = Lux.setup(rng, m)
ps_c = ComponentArrays.ComponentArray(ps)

@test Zygote.gradient(ps_c) do params
	only(first(Lux.apply(m, x, params, st)))
end isa Tuple{Nothing}

@test_throws(
"MethodError: no method matching zero(::Type{Any})",
ReverseDiff.gradient(ps_c) do params
	only(first(Lux.apply(m, x, params, st)))
end 
)

# put actual parameters into `ps_c`
m = Lux.Chain(Lux.Dense(2=>2),Lux.MaxPool((2,2)))
ps, st = Lux.setup(rng, m)
ps_c = ComponentArrays.ComponentArray(ps)

@test Zygote.gradient(ps_c) do params
	only(first(Lux.apply(m, x, params, st)))
end isa Tuple{<:ComponentArrays.ComponentVector}

@test_throws(
"MethodError: no method matching maxpool!",
ReverseDiff.gradient(ps_c) do params
	only(first(Lux.apply(m, x, params, st)))
end 
)

# Can't use Chain Rules for some reason
@test_throws(
"UndefVarError: NNlib not defined",
ReverseDiff.@grad_from_chainrules NNlib.maxpool(x::ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}, pdims::NNlib.PoolDims{2, 2, 2, 4, 2})
)
```

And the trace for
```julia
ReverseDiff.gradient(ps_c) do params
	only(first(Lux.apply(m, x, params, st)))
end 
```
```
ERROR: 
MethodError: no method matching maxpool!(::Array{ReverseDiff.TrackedReal{Float64, Float32, ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}}, 5}, ::ReverseDiff.TrackedArray{Float64, Float32, 5, Array{Float64, 5}, Array{Float32, 5}}, ::NNlib.PoolDims{3, 3, 3, 6, 3})
Closest candidates are:
  maxpool!(::AbstractArray{T, 5}, ::AbstractArray{T, 5}, ::NNlib.PoolDims; kwargs...) where T at ~/.julia/packages/NNlib/ydqxJ/src/pooling.jl:38
  maxpool!(::CUDA.CuArray{T, 3}, ::CUDA.CuArray{T, 3}, ::NNlib.PoolDims) where T<:Union{Float16, Float32, Float64} at ~/.julia/packages/NNlibCUDA/C6t0p/src/cudnn/pooling.jl:53
  maxpool!(::CUDA.CuArray{T}, ::CUDA.CuArray{T}, ::NNlib.PoolDims) where T<:Union{Float16, Float32, Float64} at ~/.julia/packages/NNlibCUDA/C6t0p/src/cudnn/pooling.jl:14
  ...
Stacktrace:
  [1] maxpool!(y::Array{ReverseDiff.TrackedReal{Float64, Float32, ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}}, 4}, x::ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}, pdims::NNlib.PoolDims{2, 2, 2, 4, 2}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ NNlib ~/.julia/packages/NNlib/ydqxJ/src/pooling.jl:73
  [2] maxpool!(y::Array{ReverseDiff.TrackedReal{Float64, Float32, ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}}, 4}, x::ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}, pdims::NNlib.PoolDims{2, 2, 2, 4, 2})
    @ NNlib ~/.julia/packages/NNlib/ydqxJ/src/pooling.jl:70
  [3] maxpool(x::ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}, pdims::NNlib.PoolDims{2, 2, 2, 4, 2}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ NNlib ~/.julia/packages/NNlib/ydqxJ/src/pooling.jl:119
  [4] maxpool(x::ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}, pdims::NNlib.PoolDims{2, 2, 2, 4, 2})
    @ NNlib ~/.julia/packages/NNlib/ydqxJ/src/pooling.jl:114
  [5] (::Lux.MaxPool{2, 4})(x::ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}, ps::ReverseDiff.TrackedArray{Float32, Float32, 1, SubArray{Float32, 1, Vector{Float32}, Tuple{UnitRange{Int64}}, true}, SubArray{Float32, 1, Vector{Float32}, Tuple{UnitRange{Int64}}, true}}, st::NamedTuple{(), Tuple{}})
    @ Lux ~/.julia/packages/Lux/mV3Ur/src/layers/conv.jl:214
  [6] apply(model::Lux.MaxPool{2, 4}, x::ReverseDiff.TrackedArray{Float64, Float32, 4, Array{Float64, 4}, Array{Float32, 4}}, ps::ReverseDiff.TrackedArray{Float32, Float32, 1, SubArray{Float32, 1, Vector{Float32}, Tuple{UnitRange{Int64}}, true}, SubArray{Float32, 1, Vector{Float32}, Tuple{UnitRange{Int64}}, true}}, st::NamedTuple{(), Tuple{}})
    @ LuxCore ~/.julia/packages/LuxCore/IgepB/src/LuxCore.jl:100
  [7] macro expansion
    @ ~/.julia/packages/Lux/mV3Ur/src/layers/containers.jl:0 [inlined]
  [8] applychain(layers::NamedTuple{(:layer_1, :layer_2), Tuple{Lux.Dense{true, typeof(identity), typeof(Lux.glorot_uniform), typeof(Lux.zeros32)}, Lux.MaxPool{2, 4}}}, x::Array{Float64, 4}, ps::ReverseDiff.TrackedArray{Float32, Float32, 1, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}}, st::NamedTuple{(:layer_1, :layer_2), Tuple{NamedTuple{(), Tuple{}}, NamedTuple{(), Tuple{}}}})
    @ Lux ~/.julia/packages/Lux/mV3Ur/src/layers/containers.jl:460
  [9] (::Lux.Chain{NamedTuple{(:layer_1, :layer_2), Tuple{Lux.Dense{true, typeof(identity), typeof(Lux.glorot_uniform), typeof(Lux.zeros32)}, Lux.MaxPool{2, 4}}}})(x::Array{Float64, 4}, ps::ReverseDiff.TrackedArray{Float32, Float32, 1, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}}, st::NamedTuple{(:layer_1, :layer_2), Tuple{NamedTuple{(), Tuple{}}, NamedTuple{(), Tuple{}}}})
    @ Lux ~/.julia/packages/Lux/mV3Ur/src/layers/containers.jl:457
 [10] apply(model::Lux.Chain{NamedTuple{(:layer_1, :layer_2), Tuple{Lux.Dense{true, typeof(identity), typeof(Lux.glorot_uniform), typeof(Lux.zeros32)}, Lux.MaxPool{2, 4}}}}, x::Array{Float64, 4}, ps::ReverseDiff.TrackedArray{Float32, Float32, 1, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}}, st::NamedTuple{(:layer_1, :layer_2), Tuple{NamedTuple{(), Tuple{}}, NamedTuple{(), Tuple{}}}})
    @ LuxCore ~/.julia/packages/LuxCore/IgepB/src/LuxCore.jl:100
 [11] (::var"#34#35")(params::ReverseDiff.TrackedArray{Float32, Float32, 1, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}})
    @ Main ./REPL[2]:2
 [12] ReverseDiff.GradientTape(f::var"#34#35", input::ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}, cfg::ReverseDiff.GradientConfig{ReverseDiff.TrackedArray{Float32, Float32, 1, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}}})
    @ ReverseDiff ~/.julia/packages/ReverseDiff/YkVxM/src/api/tape.jl:199
 [13] gradient(f::Function, input::ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}, cfg::ReverseDiff.GradientConfig{ReverseDiff.TrackedArray{Float32, Float32, 1, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}, ComponentArrays.ComponentVector{Float32, Vector{Float32}, Tuple{ComponentArrays.Axis{(layer_1 = ViewAxis(1:6, Axis(weight = ViewAxis(1:4, ShapedAxis((2, 2), NamedTuple())), bias = ViewAxis(5:6, ShapedAxis((2, 1), NamedTuple())))), layer_2 = 7:6)}}}}}) (repeats 2 times)
    @ ReverseDiff ~/.julia/packages/ReverseDiff/YkVxM/src/api/gradients.jl:22
 [14] top-level scope
    @ REPL[2]:1
 [15] top-level scope
    @ ~/.julia/packages/CUDA/N71Iw/src/initialization.jl:163
```

### PR DESCRIPTION
**Pull request by manuelbb-upb**
This is meant to fix #484 by broadening the method signatures of pooling operations.
Most importantly, the in-place methods, like `maxpool!(y, x, pdims)`, `meanpool!` and `lpnormpool` now can take arrays with different element types. This should rarely happen, but ReverseDiff ran into issues because of how it overloads `similar`.

When looking into it, I also noticed that the old implementation of `meanpool` and `lpnormpool` would have errored with non-float arrays.
That' why there are now type-computations like 
```julia
yT = if $(name == :maxpool)
    xT
elseif $(name == :meanpool)
    Base.promote_op(/, xT, Int)
else
    Base.promote_op(^, xT, Base.promote_op(/, xT, typeof(kwargs[:p])))
end
y = similar(x, yT, output_size(pdims)..., channels_out(pdims), size(x, N))
```
As this seems somewhat brittle, we could also just set `yT = Float64` if `xT` is not an `AbstractFloat` (?).

### PR Checklist

- [x] Tests are added
- [x] Documentation, if applicable

### CODE REVIEW

**Comment by ToucheSir** [comment]
In principle this looks good to me, but we should probably avoid using `promote_op` because it relies on the somewhat unreliable `_return_type`. Would ReverseDiff still work if we kept `y = similar(x, output_size(pdims)..., channels_out(pdims), size(x, N))`?

**Comment by manuelbb-upb** [comment]
Yes, it totally should. That whole `promote_op` processing is just for non-float types. Should rarely happen (maybe if someone wanted to do pre-processing on raw mnist with pooling ...)
I'll remove the type-promotions.
Regarding non-float input: Should I open a new (low-priority) issues for non-float input for discussion or
should we simply ignore it?

**Comment by ToucheSir** [comment]
I think we should ignore it for now as long as it isn't breaking Forward/ReverseDiff.jl

**Comment by manuelbb-upb** [comment]
https://github.com/FluxML/NNlib.jl/pull/485/commits/e022099b186f7942fc8f14101ea08097ae17e3a7 reverts the type prediction logic.
This still does not solve the ReverseDiff issue entirely, because `typemin` is needed for tracked reals.
I tried to change the initilization of `m` in the pooling operations to avoid `typemin` for max-pooling but always messed up some dimensions.
So I rather go with
```julia
Base.typemin(tr::Type{<:T}) where{V, D, O, T<:ReverseDiff.TrackedReal{V, D, O}} = T(typemin(V))
```
and then differentiating `maxpool` appears to work.

**Comment by CarloLucibello** [comment]
what is the status of this PR?

**Comment by manuelbb-upb** [comment]
Stagnant right now, but I can have a look tomorrow to see if it is ready for review.

File: test/pooling.jl [hunk 1]
```diff
@@ -905,6 +905,44 @@ maxpool_answer_nature = Dict(
...
+    #=
+    # This is how to test #484 with ReverseDiff, which is not a test dependency atm:
+    import ReverseDiff as RD
+    x = reshape(Float32[ 1 2; 3 4 ], (2,2,1,1))
+    @test only(maxpool(x, (2,2))) == 4
+    # @test RD.gradient(_x -> only(maxpool(_x,(2,2))), x)[:,:,1,1] == [0 0; 0 1]
+    @test only(meanpool(x, (2,2))) == 2.5
+    @test all( RD.gradient(_x -> only(meanpool(_x,(2,2))), x) .== 0.25 )
+    end
+    =#
```
Comment (darsnack) [inline test/pooling.jl:943]: I don't think there is an issue adding ReverseDiff as a test dep in order to increase our coverage? This is exactly the kind of test we need more of.

**Review by darsnack — COMMENTED** [review]
Looks ready to me. Test failures are unrelated and exist on main. Only suggestion is to add the test dependency to test ReverseDiff.

File: test/pooling.jl (hunk 1 above)
Comment (manuelbb-upb) [reply test/pooling.jl]: I added it as a test dependency and uncommented the tests.
I have defined a `typemin` for the `maxpool` line, but I can also easily make it `@test_broken`.

File: test/pooling.jl [hunk 2]
```diff
@@ -905,6 +905,42 @@ maxpool_answer_nature = Dict(
...
+    end
+    
+    # This is how to test #484 with ReverseDiff:
+    x = reshape(Float32[ 1 2; 3 4 ], (2,2,1,1))
+    @test only(maxpool(x, (2,2))) == 4
+    # define typemin, because of https://github.com/JuliaDiff/ReverseDiff.jl/issues/225
+    Base.typemin(tr::Type{<:T}) where{V, D, O, T<:RD.TrackedReal{V, D, O}} = T(typemin(V))
+    @test RD.gradient(_x -> only(maxpool(_x,(2,2))), x)[:,:,1,1] == [0 0; 0 1]
+    @test only(meanpool(x, (2,2))) == 2.5
+    @test all( RD.gradient(_x -> only(meanpool(_x,(2,2))), x) .== 0.25 )
```
Comment (ToucheSir) [inline test/pooling.jl:942]:
```suggestion
    @test all(==(0.25), RD.gradient(_x -> only(meanpool(_x,(2,2))), x))
```
I think this block of `@test`s could use some whitespace changes to match the rest of the test code.

File: test/pooling.jl [hunk 3]
```diff
@@ -905,6 +905,42 @@ maxpool_answer_nature = Dict(
...
+                @test lpnormpool!(y, x, pdims; p=1.0) isa Array{yT}
+            end
+        end
+    end
+    
+    # This is how to test #484 with ReverseDiff:
+    x = reshape(Float32[ 1 2; 3 4 ], (2,2,1,1))
+    @test only(maxpool(x, (2,2))) == 4
+    # define typemin, because of https://github.com/JuliaDiff/ReverseDiff.jl/issues/225
+    Base.typemin(tr::Type{<:T}) where{V, D, O, T<:RD.TrackedReal{V, D, O}} = T(typemin(V))
```
Comment (ToucheSir) [inline test/pooling.jl:939]:
```suggestion
    Base.typemin(tr::Type{<:T}) where{V, T<:RD.TrackedReal{V, <:Any, <:Any}} = T(typemin(V))
```

File: src/impl/pooling_direct.jl [hunk 4]
```diff
@@ -13,13 +13,13 @@ for name in (:max, :mean, :lpnorm)
     end
 
     @eval function $((Symbol("$(name)pool_direct!")))(
-        y::AbstractArray{T,5}, x::AbstractArray{T,5},
+        y::AbstractArray{T,5}, x::AbstractArray{R,5},
```
Comment (ToucheSir) [inline src/impl/pooling_direct.jl:16]:
```suggestion
        y::AbstractArray{T,5}, x::AbstractArray{<:Any,5},
```
Any type parameters not used in the function should be unnecessary

File: src/impl/pooling_direct.jl [hunk 5]
```diff
@@ -159,9 +170,9 @@ for name in (:max, :mean, :lpnorm)
     end
 
     @eval function $((Symbol("∇$(name)pool_direct!")))(
-                    dx::AbstractArray{T,5}, dy::AbstractArray{T,5},
-                    y::AbstractArray{T,5}, x::AbstractArray{T,5},
-                    pdims::PoolDims; kwargs...) where T
+                    dx::AbstractArray{DX,5}, dy::AbstractArray{DY,5},
+                    y::AbstractArray{Y,5}, x::AbstractArray{X,5},
+                    pdims::PoolDims; kwargs...) where {X, Y, DX, DY}
```
Comment (ToucheSir) [inline src/impl/pooling_direct.jl:175]: Same comment about unused type params here.

File: src/impl/pooling_direct.jl [hunk 6]
```diff
@@ -170,10 +181,10 @@ for name in (:max, :mean, :lpnorm)
     # Same story for gradients, and although this is very similar to the forward pass,
     # it's unfortunately different enough that I think we need a separate function.  :(
     @eval function $((Symbol("∇$(name)pool_direct!")))(
-                    dx::AbstractArray{T,5}, dy::AbstractArray{T,5},
-                    y::AbstractArray{T,5}, x::AbstractArray{T,5},
+                    dx::AbstractArray{DX,5}, dy::AbstractArray{DY,5},
+                    y::AbstractArray{Y,5}, x::AbstractArray{X,5},
                     pdims::PoolDims, ::Val{K}; # == kernel_size(pdims)
-                    alpha::T=T(1), beta::T=T(0), kwargs...) where {T, K}
+                    alpha=1, beta=0, kwargs...) where {DX, DY, X, Y, K}
```
Comment (ToucheSir) [inline src/impl/pooling_direct.jl:187]: Also unused type params

File: src/pooling.jl [hunk 7]
```diff
@@ -36,8 +36,8 @@ for (front_name, backend) in (
     # We only define 3d pooling primitives, we reshape lower down to get 1d and 2d pooling
     @eval begin
         function $(Symbol("$(front_name)!"))(
-                y::AbstractArray{T,5}, x::AbstractArray{T,5},
-                pdims::PoolDims; kwargs...) where {T}
+                y::AbstractArray{T,5}, x::AbstractArray{R,5},
+                pdims::PoolDims; kwargs...) where {T, R}
```
Comment (ToucheSir) [inline src/pooling.jl:40]: Just adding one comment to this file about unused type params because each function in the diff has them.

**Review by ToucheSir — COMMENTED** [review]
Some small remaining nits but otherwise LGTM too!

**Comment by manuelbb-upb** [comment]
I have removed most of the unused type parameters referring to array element types. 
`<:Any` is functionally equivalent to how it was before.
In practice, we can expect arrays of Floats, `meanpool` and `lpnormpool` will still fail on integer arrays.

**Review by ToucheSir — APPROVED** [review]
Great! I think this just needs a rebase to resolve the conflicts.

**Comment by manuelbb-upb** [comment]
Sorry that it took so long. Did the rebase.

**Comment by ToucheSir** [comment]
Thanks!

**Pull request merged**